src/main/scala/com.snowplowanalytics.sauna/responders/hipchat/SendRoomNotificationResponder.scala

```diff
@@ -0,0 +1,238 @@
+
+  /**
+   * A Sauna command envelope.
+   *
+   * @param commandId   A unique identifier (uuid4) for the command.
+   * @param whenCreated The command's creation date.
+   * @param execution   TODO: define what this is
+   * @param tags        TODO: define what this is
+   */
+  case class CommandEnvelope(
+    commandId: String,
+    whenCreated: LocalDateTime,
+    execution: ExecutionParams,
+    tags: Map[String, String]
+  )
+  implicit val commandEnvelopeReads: Reads[CommandEnvelope] = Json.reads[CommandEnvelope]
+
+  /**
+   * Constructs a [[Props]] for a [[SendRoomNotificationResponder]] actor.
+   *
+   * @param hipchat The HipChat API wrapper.
+   * @param logger  A logger actor.
+   * @return [[Props]] for the new actor.
+   */
+  def props(hipchat: Hipchat, logger: ActorRef): Props =
+    Props(new SendRoomNotificationResponder(hipchat, logger))
+
+  /**
+   * Attempts to extract a Sauna command from a [[JsValue]].
+   *
+   * @param json The [[JsValue]] to extract a command from.
+   * @tparam T The type of the command's data.
+   * @return Right containing a tuple of the command's envelope and data
+   *         if the extraction was successful, Left containing an error message
+   *         otherwise.
+   */
+  def extractCommand[T](json: JsValue)(implicit tReads: Reads[T]): Either[String, (CommandEnvelope, T)] = {
+    json.validate[SelfDescribing] match {
+      case JsSuccess(selfDescribing, _) =>
+        selfDescribing.data.validate[SaunaCommand] match {
+          case JsSuccess(command, _) =>
+            command.envelope.data.validate[CommandEnvelope] match {
+              case JsSuccess(envelope, _) =>
+                command.command.data.validate[T] match {
+                  case JsSuccess(data, _) =>
+                    Right((envelope, data))
+                  case JsError(error) =>
+                    Left(s"Encountered an issue while parsing Sauna command data: $error")
+                }
+              case JsError(error) =>
+                Left(s"Encountered an issue while parsing Sauna command envelope: $error")
+            }
+          case JsError(error) =>
+            Left(s"Encountered an issue while parsing Sauna command: $error")
+        }
+      case JsError(error) =>
+        Left(s"Encountered an issue while parsing self-describing JSON: $error")
+    }
+  }
+
+  /**
+   * Processes a Sauna command envelope.
+   *
+   * @param envelope A Sauna command envelope.
+   * @return None if the envelope was successfully processed
+   *         and the command's data can be executed,
+   *         Some containing an error message otherwise.
+   */
+  def processEnvelope(envelope: CommandEnvelope): Option[String] = {
+    envelope.execution.timeToLive match {
+      case Some(ms) =>
+        val commandLife = envelope.whenCreated.until(LocalDateTime.now(), ChronoUnit.MILLIS)
+        if (commandLife > ms)
+          return Some(s"Command has expired: time to live is $ms but $commandLife has passed")
```

Comment:
I don't think there's a way to avoid `return` with this particular control flow (omitting it simply returns a `None` later on).